openpdf/src/main/java/com/lowagie/text/pdf/CMapAwareDocumentFont.java

```diff
@@ -216,32 +216,87 @@ private String decodeSingleCID(byte[] bytes, int offset, int len) {
 		throw new Error("Multi-byte glyphs not implemented yet");
 	}
 
-	/**
-	 * Decodes a string of bytes (encoded in the font's encoding) into a unicode
-	 * string This will use the ToUnicode map of the font, if available,
-	 * otherwise it uses the font's encoding
-	 * 
-	 * @param cidbytes
-	 *            the bytes that need to be decoded
-	 * @return the unicode String that results from decoding
-	 * @since 2.1.7
-	 */
-	public String decode(byte[] cidbytes, final int offset, final int len) {
-		StringBuffer sb = new StringBuffer(); // it's a shame we can't make this
-												// StringBuilder
-		for (int i = offset; i < offset + len; i++) {
-			String rslt = decodeSingleCID(cidbytes, i, 1);
-			if (rslt == null && i + 1 < offset + len) {
-				rslt = decodeSingleCID(cidbytes, i, 2);
-				i++;
-			}
-			if (rslt != null) {
-				sb.append(rslt);
-			}
-		}
+    /**
+     * @return true if this font has unicode information available.
+     */
+    public boolean hasUnicodeCMAP() {
+        return toUnicodeCmap != null;
+    }
 
-		return sb.toString();
-	}
+    /**
+     * Decodes a string of bytes (encoded in the font's encoding) into a unicode string. This will
+     * use the ToUnicode map of the font, if available, otherwise it uses the font's encoding
+     *
+     * @param cidbytes
+     *            the bytes that need to be decoded
+     * @return the unicode String that results from decoding
+     * @since 2.1.7
+     */
+    public String decode(byte[] cidbytes,
+                         final int offset,
+                         final int len) {
+        StringBuffer sb = new StringBuffer(); // it's a shame we can't make this
```

Comment:
Why can't we use StringBuilder? Is this an old comment?

Reply:
I should have put in a longer comment. As you can see below, the explanation may have gotten away from me a bit.

PdfString strings sometimes contain PDF code points (values 0-256) and not proper UTF-8 characters (or indeed characters in any any defined encoding). StringBuilder makes "real" strings, i.e. strings in a defined encoding. The unicode characters corresponding to values from 128-255 take multiple byes in UTF-8 (and a few even do so in even UTF-16). StringBuffer is an old enough class that we can abuse encodings as in the original code, and manage not to end up with "too many bytes" (or not enough byes) at the end, because it will let us plot bytes into a string without caring if the result is properly encoded.

Using Strings keeps the API compatible, but at a cost to the internal implementation.

I considered a "real" fix, but it would have involved significant changes to the PdfString implementation and API. That seems like a lot of disruption to users; worse, it would involve touching most of the code that uses PdfString, which seemed like a way to create more bugs than I would solve. The Stream parsing code is low-level and tricky, and I'm also not confident that I really understand all the encodings and what the best strategies are for each.

The original code doesn't pretend to be completely correct, it glosses over and omits various cases; I added some additional support for Unicode fonts, but there are sure to be more problems to fix over time.

Reply:
Nice explanation. Thanks.